play-json/shared/src/main/scala/play/api/libs/json/JsValue.scala

```diff
@@ -190,14 +191,20 @@ case class JsObject(
     merge(this, other)
   }
 
-  override def equals(other: Any): Boolean = other match {
-    case that @ JsObject(_) => (that.canEqual(this)) && fieldSet == that.fieldSet
-    case _                  => false
+  override def equals(other: Any): Boolean = {
+    other match {
+      case o: AnyRef if this.eq(o) =>
```

Comment:
How could this happen?

Reply:
Good question! As an illustrative example, consider a json text that looks like `[[{"k":0}], [{"k":0}],[{"k":0}],[{"k":0}],...]` whose length approaches the jvm's total heap size. `Json.parse(file)` would throw `OutOfMemoryError`, but is it _possible_ to represent such a text as a `JsValue`?

Currently, jackson-core will return the same `"k"` `String` instance for all keys in the example, but that's only a small part of the eventual `JsValue` memory footprint. It's possible to apply the same technique to the other duplicated values as well.

In weepickle terms, a minimal implementation could look like:
```scala
val jsValue = FromJson(inputStream).transform(new DdVisitor(PlayJson))

class DdVisitor[T, J](underlying: Visitor[T, J])
 (implicit cache: mutable.Map[Any, Any] = mutable.Map.empty)
 extends Visitor.Delegate[T, J](underlying)
   with RecursiveDelegate[T, J] {

 override def asRecursiveDelegate[A, B](visitor: Visitor[A, B]) =
   new DdVisitor(
     visitor.map { jsValue =>
       cache
         .getOrElseUpdate(jsValue, jsValue)
         .asInstanceOf[B]
     }
   )
}
```

As we walk the input depth-first, every `JsValue` node is run through a `Map[JsValue, JsValue]` to return a canonical instance. Each `{"k":0}` would result in a cache hit and return a single canonical instance.

Back to your question, this line would be hit when we attempt to deduplicate `[{"k":0}]`. In order to check if two `JsArray` are identical, it needs to test that all elements pass equals(). Since the `{"k":0}` `JsObject` that's present in both arrays would have already been canonicalized to the same instance, `this eq other` would be true for that comparison.

I wish this were a purely hypothetical scenario. In practice, we talk to some APIs that return duplication-heavy JSON texts that have OOME'd our apps before. My favorite example was a 5.4 GB `JsObject` in heap whose UTF-8 input compressed down to 1.7 MB. Those units are not typos. That's ~3,000x memory amplification.

One of the things we like about play-json is that it gives us the flexibility to mitigate these problems by extending the JsValue classes. Applying deduplication brought the 5.4 GB JsValue down to a 205 MB JsValue. Bringing our [own Map/Seq](https://github.com/playframework/play-json/pull/675#discussion_r780752661) and doing some unboxing got the same input down to a 23 MB JsValue. Throughput was also faster due to eliminated GC pressure. Win-win. (We also logged an issue with the service responding with the extra data to please stop sending us info we don't need, but I was on-call and wanted to stop getting paged about this.)

Some APIs are highly normalized and duplicate-free, but in practice, I suspect many of them contain significant duplication. Using the few thousand test json files in our test double service as a proxy for the real-world data we deal with, I found that 71% of the nodes were present somewhere else in the same json text. Duplication wasn't present only in primitives either. I found significant duplication up to height=5 of the trees. I'm very curious if we just deal with bad REST APIs or if this is more widespread.

Anyway, over typical usage this line won't hit, although it's quite cheap, especially compared to the previous implementation that constructed two copies of `underlying` on equals()--one of them a `HashSet`! If needed, we could remove it and get by with the underlying Map performing the `eq` check, but while I was in here, I saw the opportunity to short-circuit, tried it while benchmarking, and left it.

Thanks for the early feedback on this draft! Code is actually ready for review, but I didn't want to send reviewer notifications until I'd written the PR description. Now that I've typed all this out, maybe I'll just skip it. :)

Reply:
K, would be interested to add a benchmark after.

Reply:
Before:
```
[info] Benchmark                    (size)   Mode  Cnt  Score   Error   Units
[info] JsObjectBench.oEq                 1  thrpt    5  5.221 ± 0.194  ops/us
[info] JsObjectBench.oEq                15  thrpt    5  0.292 ± 0.007  ops/us
[info] JsObjectBench.oEquals             1  thrpt    5  5.050 ± 0.124  ops/us
[info] JsObjectBench.oEquals            15  thrpt    5  0.279 ± 0.016  ops/us
[info] JsObjectBench.oHashCodeWarm       1  thrpt    5  8.608 ± 0.524  ops/us
[info] JsObjectBench.oHashCodeWarm      15  thrpt    5  0.513 ± 0.012  ops/us
```

After:
```
[info] Benchmark                    (size)   Mode  Cnt    Score    Error   Units
[info] JsObjectBench.oEq                 1  thrpt    5  420.943 ± 13.411  ops/us
[info] JsObjectBench.oEq                15  thrpt    5  418.040 ± 19.643  ops/us
[info] JsObjectBench.oEquals             1  thrpt    5  314.461 ± 14.557  ops/us
[info] JsObjectBench.oEquals            15  thrpt    5  310.431 ± 25.338  ops/us
[info] JsObjectBench.oHashCodeWarm       1  thrpt    5   20.217 ±  1.559  ops/us
[info] JsObjectBench.oHashCodeWarm      15  thrpt    5    3.830 ±  0.311  ops/us
```

Reply:
Longer run, higher precision, prettier graphs:
https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/htmldoug/6ebb5f3f169e35535a99e2d198c9153d/raw/f8f1a3b905de98829e607ad41fe8ffa936ffd6c8/report-before.json,https://gist.githubusercontent.com/htmldoug/6ebb5f3f169e35535a99e2d198c9153d/raw/f8f1a3b905de98829e607ad41fe8ffa936ffd6c8/report-after.json